### PR TITLE
daemon/linux: Set console size on creation

### PR DESCRIPTION
**Pull request by vvoland**
**- What I did**
On Linux the daemon was not respecting the HostConfig.ConsoleSize property and relied on cli initializing the tty size after the container was created. This caused a delay between container creation and the tty actually being resized.

So for example:
`docker run -it --rm ubuntu sh -c 'tput cols; tput lines'`
would print the default tty size

but:
`docker run -it --rm ubuntu sh -c 'sleep 1; tput cols; tput lines'`
would print the expected size

This is also a small change to the api description, because HostConfig.ConsoleSize is no longer Windows-only

Relates to
[cli#3554](https://github.com/docker/cli/issues/3554)
[cli#3572](https://github.com/docker/cli/pull/3572#issuecomment-1112069795)

**- How I did it**
Apply the HostConfig.ConsoleSize to the Spec just like the Windows code does.

**- How to verify it**
Terminal size should be changed immediately when starting container, can verify by resizing the terminal window from default 80x24 and checking if running the below prints the expected size:
`docker run -it --rm ubuntu sh -c 'tput cols; tput lines'`

**- Description for the changelog**
Container's tty size is set immediately on creation

**- A picture of a cute animal (not mandatory but encouraged)**
cute Duduś

### CODE REVIEW

**Comment by AkihiroSuda** [comment]
Can we have a test in https://github.com/moby/moby/tree/master/integration/container ?

**Comment by thaJeztah** [comment]
CI failure is unrelated (known flaky test), so can be ignored

**Comment by thaJeztah** [comment]
A test for this would indeed be good; other changes;


can you add a blurb to https://github.com/moby/moby/blob/master/docs/api/version-history.md to mention that the ConsoleSize is now also accepted on Linux

If we want to do it "correct", technically we should reset the value on older API versions
 
so both the client https://github.com/moby/moby/blob/bf04690bbc5e4226e3f8ea3ce4c8251766b84ad3/client/container_create.go#L23-L33

and the API endpoint;
https://github.com/moby/moby/blob/f1dd6bf84e28930e1ccd903361f9284fb22d3b8a/api/server/router/container/container_routes.go#L493

should ignore the value (resetting them) on Linux

For this one, I'm _slightly_ on the fence, as (I think?) the client will still do a resize on attach (or can we get rid of the resize loop on the client for API >= 1.42 (if that's the case, we should consider ignoring it on older versions to prevent inconsistent behaviour when connecting a new CLI with an older daemon)


I notice that on the CLI side, we probably need some more changes as well; it looks like we only set the size on `runContainer()`; https://github.com/docker/cli/blob/a221771e4ac106909ec60e94f0fd5ede9ae592fd/cli/command/container/run.go#L121-L124

But don't do it on `createContainer()` https://github.com/docker/cli/blob/5c769c40bee8001f33666588ca8fd1b8eb5c12c5/cli/command/container/create.go#L194

(wondering if it should be moved to `createContainer()` so that it's also done on `docker create` / `docker container create` - although, given, we likely still would need to do a resize on `docker attach` in either case.

File: daemon/oci_linux.go [hunk 1]
```diff
@@ -989,6 +989,17 @@ func WithSysctls(c *container.Container) coci.SpecOpts {
 	}
 }
 
+// WithConsoleSize sets the initial console size
+func WithConsoleSize(c *container.Container) coci.SpecOpts {
```
Comment (thaJeztah) [inline daemon/oci_linux.go:993]: At some point we should move these to a non-"linux" file, as most of these could be equally used for Windows (not needed for this PR, we can probably just rename the file at some point and update the Windows code)

**Comment by vvoland** [comment]
Added test, version-history blurb and API checks.

As for the CLI changes:

> I notice that on the CLI side, we probably need some more changes as well; it looks like we only set the size on runContainer(); https://github.com/docker/cli/blob/a221771e4ac106909ec60e94f0fd5ede9ae592fd/cli/command/container/run.go#L121-L124
But don't do it on createContainer() https://github.com/docker/cli/blob/5c769c40bee8001f33666588ca8fd1b8eb5c12c5/cli/command/container/create.go#L194
(wondering if it should be moved to createContainer() so that it's also done on docker create / docker container create - although, given, we likely still would need to do a resize on docker attach in either case.

Makes sense to me, I will make cli PR soon.


> For this one, I'm slightly on the fence, as (I think?) the client will still do a resize on attach (or can we get rid of the resize loop on the client for API >= 1.42 (if that's the case, we should consider ignoring it on older versions to prevent inconsistent behaviour when connecting a new CLI with an older daemon)

Shouldn't we still leave the initial tty resize behaviour before entering the "wait for tty size change" loop though? 
In theory there is a some window of time between container creation and the MonitorTtySize loop during which the cli's terminal size could change.
Probably it's not a very likely scenario, but since we still need to do this for exec/attach/start then maybe it's not worth to add extra API checks just to save us from doing it for run?

File: daemon/oci_linux.go (hunk 1 above)
Comment (vvoland) [reply daemon/oci_linux.go]: Moved this to a non-linux file. The Windows code doesn't set the spec with the oci.SpecOpts, so for now it can't really share this function with linux code.

**Comment by thaJeztah** [comment]
> Probably it's not a very likely scenario, but since we still need to do this for exec/attach/start then maybe it's not worth to add extra API checks just to save us from doing it for run?

Yeah, good point; perhaps we can't get rid of it (was mostly thinking that the retry loop could in some cases lead to the error being printed, which (with this PR implemented) could possibly lead to unneeded noise) 🤔 

Anyway; those are things we can discuss on the CLI repository; I think the "move it to `createContainer()`" change at least is something we should look at (in the CLI), so that `docker container create ... && docker container start` at least has the initial size set correctly, and not only `docker run`.

File: daemon/oci_opts.go [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+	specs "github.com/opencontainers/runtime-spec/specs-go"
+)
+
+// WithConsoleSize sets the initial console size
+func WithConsoleSize(c *container.Container) coci.SpecOpts {
+	return func(ctx context.Context, _ coci.Client, _ *containers.Container, s *coci.Spec) error {
+		s.Process.ConsoleSize = &specs.Box{
+			Height: c.HostConfig.ConsoleSize[0],
+			Width:  c.HostConfig.ConsoleSize[1],
+		}
```
Comment (thaJeztah) [inline daemon/oci_opts.go:18]: To be on the safe side, perhaps we should add a check here, so that we don't override the default (`nil`) value if there's no size to be set;

```suggestion
		if c.HostConfig.ConsoleSize[0] > 0 || c.HostConfig.ConsoleSize[1] > 0 {
			s.Process.ConsoleSize = &specs.Box{
				Height: c.HostConfig.ConsoleSize[0],
				Width:  c.HostConfig.ConsoleSize[1],
			}
		}
```

Comment (thaJeztah) [reply daemon/oci_opts.go:18]: FWIW; we should start looking at using the containerd options for these things, but they use a different signature (all options in our local package basically take a full Config, so that requires quote some refactoring); e.g. in this case, we'd have to replace with https://github.com/moby/moby/blob/4cac624fefefb7bb6640f205587d0dc0a9651b27/vendor/github.com/containerd/containerd/oci/spec_opts.go#L246-L258


But that's definitely out of scope for this PR

File: integration/container/run_test.go [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+func TestConsoleSize(t *testing.T) {
+	skip.If(t, versions.LessThan(testEnv.DaemonAPIVersion(), "1.42"), "skip test from new feature")
+
+	defer setupTest(t)()
+	client := testEnv.APIClient()
+	ctx := context.Background()
+
+	cID := container.Run(ctx, t, client,
+		container.WithTty(true),
+		container.WithImage("debian:bullseye-slim"),
```
Comment (thaJeztah) [inline integration/container/run_test.go:28]: Was wondering why not busybox, but looks like that doesn't support `tput` 🤷‍♂️

File: integration/container/run_test.go [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/versions"
+	"github.com/docker/docker/integration/internal/container"
+	"gotest.tools/v3/assert"
+	"gotest.tools/v3/poll"
+	"gotest.tools/v3/skip"
+)
+
+func TestConsoleSize(t *testing.T) {
```
Comment (thaJeztah) [inline integration/container/run_test.go:19]: Should this test be in the `run_linux_test.go` file? (as Windows won't run `debian`)

File: integration/container/run_test.go [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+	ctx := context.Background()
+
+	cID := container.Run(ctx, t, client,
+		container.WithTty(true),
+		container.WithImage("debian:bullseye-slim"),
+		container.WithCmd("sh", "-c", "tput lines; tput cols"),
+		container.WithConsoleSize(57, 123),
+	)
+
+	poll.WaitOn(t, container.IsInState(ctx, client, cID, "exited"), poll.WithDelay(100*time.Millisecond))
```
Comment (thaJeztah) [inline integration/container/run_test.go:33]: Nit: looks like we have a `container.IsStopped()` check, so this could be;

```suggestion
	poll.WaitOn(t, container.IsStopped(ctx, client, cID), poll.WithDelay(100*time.Millisecond))
```

File: integration/container/run_test.go [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+	out, err := client.ContainerLogs(ctx, cID, types.ContainerLogsOptions{ShowStdout: true})
+	assert.NilError(t, err)
+	defer out.Close()
```
Comment (thaJeztah) [inline integration/container/run_test.go:37]: It's a bit unfortunate that we currently don't have a way to return the output from `container.Run()` (as we have for `container.Exec()`). Perhaps something we should look into (we could have a `WithOutputBuffer()` (e.g.) function that we can pass a buffer).

Also not something for this PR though; just thinking out loud here 😅

**Review by thaJeztah — CHANGES_REQUESTED** [review]
Some suggestions (and some rambling from my side; feel free to ignore those 😂)

**Comment by thaJeztah** [comment]
Looks like some tests are failing (I'm seeing some panics / nil-pointer exceptions); could be badly written tests that don't set up the right things, but should be looked at.

File: integration/container/run_test.go (hunk 3 above)
Comment (vvoland) [reply integration/container/run_test.go:28]: Yeah, unfortunately it doesn't. However some other tests also use this image, so I hope that it doesn't actually make the test suite heavier :)

Comment (thaJeztah) [reply integration/container/run_test.go:28]: It's part of the "frozen images" which are only pulled once, and loaded into test-daemons, so it should not make a huge difference https://github.com/moby/moby/blob/c9e19a2aa182e06ebe1653a2c3af02563acf9e36/Dockerfile#L88-L93


We should look at only conditionally loading some of those though, as I think we currently always load all of them in every test daemon

Comment (vvoland) [reply integration/container/run_test.go:28]: I think it's possible to use busybox by using `stty size` instead of tput:
`docker run -it busybox sh -c 'stty size'`
which returns height and width like this (it even follows our favorite order! :smile:):
`51 234`

The only caveat is that with the previous behaviour (no console size being set on creation) stty does not output the size, but:
`stty: standard input`

File: docs/api/version-history.md [hunk 7]
```diff
@@ -94,6 +94,11 @@ keywords: "API, Docker, rcli, REST, documentation"
   actually supported. API versions before v1.42 continue to ignore these parameters
   and default to attaching to all streams. To preserve the pre-v1.42 behavior,
   set all three query parameters (`?stdin=1,stdout=1,stderr=1`).
+* `/containers/{name}/attach/ws` endpoint only attach to configured streams
```
Comment (thaJeztah) [inline docs/api/version-history.md:97]: Oh! Looks like a bad rebase; this bullet is unrelated

**Review by thaJeztah — APPROVED** [review]
changes LGTM, but looks like a small fix is needed in the API history markdown (otherwise good to go)

File: docs/api/version-history.md (hunk 7 above)
Comment (vvoland) [reply docs/api/version-history.md]: Whops, my bad. Fixed!

**Review by thaJeztah — APPROVED** [review]
LGTM

File: api/server/router/container/container_routes.go [hunk 8]
```diff
@@ -517,6 +518,11 @@ func (s *containerRouter) postContainersCreate(ctx context.Context, w http.Respo
 		hostConfig.KernelMemory = 0
 	}
 
+	if hostConfig != nil && runtime.GOOS == "linux" && versions.LessThan(version, "1.42") {
```
Comment (ndeloof) [inline api/server/router/container/container_routes.go:521]: version 1.42 introduces support for this, but doesn't make it required. you updated client to enforce a value, but an  alternative client (let's say: docker-py) could call this API with a non-set ConsoleSize
 
We could just guard this by `len(hostConfig.ConsoleSize) < 0`, so that this basically is "optional"

File: client/container_create.go [hunk 9]
```diff
@@ -27,11 +27,18 @@ func (cli *Client) ContainerCreate(ctx context.Context, config *container.Config
...
+	clientVersion := cli.ClientVersion()
+
 	// When using API 1.24 and under, the client is responsible for removing the container
-	if hostConfig != nil && versions.LessThan(cli.ClientVersion(), "1.25") {
+	if hostConfig != nil && versions.LessThan(clientVersion, "1.25") {
 		hostConfig.AutoRemove = false
 	}
 
+	// When using API under 1.42, the Linux daemon doesn't respect the ConsoleSize
+	if hostConfig != nil && platform != nil && platform.OS == "linux" && versions.LessThan(clientVersion, "1.42") {
```
Comment (ndeloof) [inline client/container_create.go:38]: by my comment on endpoint, I don't think we should have the client manage default values to be set.

**Comment by thaJeztah** [comment]
CI Failure is unrelated and can be ignored; `TestAPISwarmNodeDrainPause` is a known flaky test (https://github.com/moby/moby/issues/23516)

```
=== RUN   TestDockerSwarmSuite/TestAPISwarmNodeDrainPause
    docker_api_swarm_node_test.go:81: [d6dc1f01f0219] joining swarm manager [d6ff3be5b0fb0]@0.0.0.0:2477, swarm listen addr 0.0.0.0:2478
    docker_api_swarm_node_test.go:107: timeout hit after 1m0s: output: "1a0cd918e94e\n4bfa52f81731\n"
    --- FAIL: TestDockerSwarmSuite/TestAPISwarmNodeDrainPause (73.39s)
```

File: daemon/oci_opts.go [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+	"github.com/containerd/containerd/containers"
+	coci "github.com/containerd/containerd/oci"
+	"github.com/docker/docker/container"
+	specs "github.com/opencontainers/runtime-spec/specs-go"
+)
+
+// WithConsoleSize sets the initial console size
+func WithConsoleSize(c *container.Container) coci.SpecOpts {
+	return func(ctx context.Context, _ coci.Client, _ *containers.Container, s *coci.Spec) error {
+		if c.HostConfig.ConsoleSize[0] > 0 || c.HostConfig.ConsoleSize[1] > 0 {
```
Comment (ndeloof) [inline daemon/oci_opts.go:15]: maybe add a fool-guard check that `len(HostConfig.ConsoleSize) == 2` here
(That's a bit sad we don't have a well defined struct here)

**Review by ndeloof — CHANGES_REQUESTED** [review]
some minor consideration regarding management of default value `{0,0}` based on API version, otherwise LGTM. Thanks!

File: api/server/router/container/container_routes.go (hunk 8 above)
Comment (thaJeztah) [reply api/server/router/container/container_routes.go:521]: Unfortunately, `ConsoleSize` is an array (fixed size), not a slice, so we can't set it to `nil` `[2]uint{0, 0}` is the default value for a `[2]uint`, so what this does is "reset the value as if it wasn't set" on older API version running on Linux.

File: client/container_create.go (hunk 9 above)
Comment (thaJeztah) [reply client/container_create.go:38]: See my other comment; effectively this is "resetting" the value back to the default for the type https://github.com/moby/moby/pull/43593/files#r876720241

(I guess `[2]int{}` would've worked as well if you think that's clearer though)

File: daemon/oci_opts.go (hunk 10 above)
Comment (vvoland) [reply daemon/oci_opts.go:15]: Is it really needed given that ConsoleSize has type [2]uint? One could assume that it's len will always be 2, but maybe it's some Golang's gotcha that I'm not aware of?

**Review by ndeloof — APPROVED** [review]
> ConsoleSize is an array (fixed size), not a slice, so we can't set it to nil
got it!

File: daemon/oci_opts.go (hunk 10 above)
Comment (thaJeztah) [reply daemon/oci_opts.go:15]: Double-checked (as arrays are not very commonly used in Go (nor recommended for most uses)), but yeah `[2]int` is a type, so always has 2 as len; https://go.dev/play/p/UxzMidJXmnb

trying to assign a `[]int` slice, even with len(2) will fail;

```go
var foo [2]int
foo = []int{1, 2}
```

That will produce:

```
cannot use []int{…} (value of type []int) as type [2]int in assignment
```


Good "catch" though! (definitely not a useless comment @ndeloof 👍 )

**Comment by thaJeztah** [comment]
Let me merge this one 👍 

Thanks for review!
